### PR TITLE
FXIOS-3065 - Inactive Tabs 

### PR DESCRIPTION
**Pull request by nbhasin2**
This feature enables tabs that have been inactive for 14days and more to be automatically transferred to the inactive tab tray as long as the feature is enabled. 

Currently, this feature is only being enabled for Dev and Beta builds. 

What's working: 
- Everything should be working 
- Swipe to delete an inactive tab
- Close all inactive tabs

What's missing: 
- [ ] Minor padding issue on iPhone 
- [ ] Cleanup
- Nimbus setup 
- CFR setup 

Since we have multiple configurations, it would be nice to test Inactive Tabs, Group Tabs and regular tabs.
a) Inactive Tabs + Group Tabs only
b) Inactive Tabs + Group Tabs + Regular Tabs
c) Inactive Tabs + Regular Tabs only

Below are some screenshot from the work: 

<table>
  <tr>
    <th>iPad</th>
    <th>iPhone</th>
  </tr>
  <tr>
    <td>
Screen Shot 2022-02-11 at 3 14 24 PM</td>
    <td>Screen Shot 2022-02-11 at 3 04 48 PM</td>
  </tr>
</table>

### CODE REVIEW

**Comment by nbhasin2** [comment]
# Request for data collection review form

**All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.**

1) What questions will you answer with this data?
How do our users make use of the inactive tabs feature and how often it is being used

2) Why does Mozilla need to answer these questions?  Are there benefits for users? Do we need this information to address product or business requirements? 
So the business can determine if users are using the inactive tabs

3) What alternative methods did you consider to answer these questions? Why were they not sufficient?
This is a new feature and there's no existing metrics to measure it.

4) Can current instrumentation answer these questions?
Not really as this feature is just added so current instrumentation cannot determine the preference change for this new setting.

5) List all proposed measurements and indicate the category of data collection for each measurement, using the [Firefox data collection categories](https://wiki.mozilla.org/Data_Collection) found on the Mozilla wiki.   

**Note that the data steward reviewing your request will characterize your data collection based on the highest (and most sensitive) category.**

<table>
<tr>
<td>Measurement Description</td>
<td>Data Collection Category</td>
<td>Tracking Bug #</td>
</tr>
<tr>
<td><span data-code-marker=" ">toggle_inactive_tab_tray which counts the number of times user toggle (expand or collapse) inactive tab section.</span></td>
<td>Category 2</td>
<td>https://github.com/mozilla-mobile/firefox-ios/pull/9993</td>
</tr>
<tr>
<td><span data-code-marker=" ">open_inactive_tab</span> which is recorded when a user taps on an inactive tab</td>
<td>Category 2</td>
<td>https://github.com/mozilla-mobile/firefox-ios/pull/9993</td>
</tr>
<tr>
<td><span data-code-marker=" "><span data-code-marker="+">inactive_tabs_close_all_btn is recorded when a user taps on close all inactive tabs button<br /></span></span></td>
<td>Category 2</td>
<td>https://github.com/mozilla-mobile/firefox-ios/pull/9993</td>
</tr>
<tr>
<td><span data-code-marker=" "><span data-code-marker="+">inactive_tab_swipe_close is recorded when a user swipes to close an inactive tab<br /></span></span></td>
<td>Category 2</td>
<td>https://github.com/mozilla-mobile/firefox-ios/pull/9993</td>
</tr>
<tr>
<td><span data-code-marker=" "><span data-code-marker="+">inactive_tab_shown is recorded when a user is shown inactive tab view</span></span></td>
<td>Category 2</td>
<td>https://github.com/mozilla-mobile/firefox-ios/pull/9993</td>
</tr>
</table>

6) Please provide a link to the documentation for this data collection which describes the ultimate data set in a public, complete, and accurate way.

Link to description [here](https://github.com/mozilla-mobile/firefox-ios/blob/d58ba8b691453769f80d9e9b652a4f58a5f65977/Client/metrics.yaml#L1892)

7) How long will this data be collected?  Choose one of the following:
Events will expire on 2022-08-01

8) What populations will you measure?
All users in beta or who get to use the feature considering some might be in an experiment.

9) If this data collection is default on, what is the opt-out mechanism for users?
In Firefox iOS there is a preference for telemetry labeled "Send Usage Data" that will be used as the opt-out mechanism.

10) Please provide a general description of how you will analyze this data.
We will look at how the users interact with inactive tabs component 

11) Where do you intend to share the results of your analysis?
Mozilla internal channels with our PM

12) Is there a third-party tool (i.e. not Telemetry) that you are proposing to use for this data collection? If so:
No 3rd party tool

**Comment by dnarcese** [comment]
r+
# Data Review Form

_1) Is there or will there be **documentation** that describes the schema for the ultimate data set in a public, complete, and accurate way?_ 

Yes, telemetry is documented at https://dictionary.telemetry.mozilla.org/apps/firefox_ios

_2) Is there a control mechanism that allows the user to turn the data collection on and off?_ 

Yes, Firefox iOS provides data controls in the app settings

_3) If the request is for permanent data collection, is there someone who will monitor the data over time?_

This should be re-reviewed w/ Ashley Thomas every 6mo

_4) Using the **[category system of data types](https://wiki.mozilla.org/Firefox/Data_Collection)** on the Mozilla wiki, what collection type of data do the requested measurements fall under?_

Type 2

_5) Is the data collection request for default-on or default-off?_

Default on

_6) Does the instrumentation include the addition of **any *new* identifiers** (whether anonymous or otherwise; e.g., username, random IDs, etc.  See the appendix for more details)?_

No

_7) Is the data collection covered by the existing Firefox privacy notice?_ 

Yes

_8) Does the data collection use a third-party collection tool?_ 

No

File: Client/Frontend/Browser/Tabs/InactiveTabCell.swift [hunk 1]
```diff
@@ -53,19 +68,36 @@ class InactiveTabCell: UICollectionViewCell, NotificationThemeable, UITableViewD
...
         fatalError("init(coder:) has not been implemented")
     }
-
+    
+    private func setupNotifcation() {
+        NotificationCenter.default.addObserver(self, selector: #selector(applyTheme), name: .DisplayThemeChanged, object: nil)
+    }
+    
     private func setupConstraints() {
+        containerView.snp.makeConstraints { make in
```
Comment (nbhasin2) [inline Client/Frontend/Browser/Tabs/InactiveTabCell.swift:87]: personal nit: move away from snapkit

File: Client/Frontend/Browser/Tabs/InactiveTabCell.swift [hunk 2]
```diff
@@ -53,19 +68,36 @@ class InactiveTabCell: UICollectionViewCell, NotificationThemeable, UITableViewD
...
+            make.top.equalToSuperview().offset(5)
+            make.bottom.equalToSuperview().offset(-5)
+            make.leading.equalToSuperview()
+            make.trailing.equalToSuperview()
+        }
+
         tableView.snp.makeConstraints { make in
-            make.edges.equalToSuperview()
```
Comment (nbhasin2) [inline Client/Frontend/Browser/Tabs/InactiveTabCell.swift:67]: personal nit: move away from snapkit

File: Client/Frontend/Widgets/CellWithRoundedButton.swift [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+class CellWithRoundedButton: UITableViewCell, NotificationThemeable {
+    var selectedView: UIView = {
+        let view = UIView()
+        view.backgroundColor = UIColor.theme.tableView.selectedBackground
+        return view
+    }()
+    
+    private lazy var roundedButton: UIButton = {
```
Comment (nbhasin2) [inline Client/Frontend/Widgets/CellWithRoundedButton.swift:21]: personal nit: would be better to use the uibutton builder here

File: Client/Frontend/Widgets/CellWithRoundedButton.swift [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+
+struct CellWithRoundedButtonUX {
+    static let ImageSize: CGFloat = 29
+    static let ImageCornerRadius: CGFloat = 6
+    static let HorizontalMargin: CGFloat = 16
+}
+
+class CellWithRoundedButton: UITableViewCell, NotificationThemeable {
+    var selectedView: UIView = {
```
Comment (nbhasin2) [inline Client/Frontend/Widgets/CellWithRoundedButton.swift:15]: personal nit: would be better to use the uiview builder

File: Client/Frontend/Widgets/CellWithRoundedButton.swift [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    var shouldLeftAlignTitle = false
+    var customization: OneLineTableViewCustomization = .regular
+    func initialViewSetup() {
+        separatorInset = UIEdgeInsets(top: 0, left: 0, bottom: 0, right: 0)
+        self.selectionStyle = .default
+        
+        contentView.addSubview(roundedButton)
+        roundedButton.titleEdgeInsets = UIEdgeInsets(top: 0, left: 30, bottom: 0, right: 0)
+        
+        roundedButton.snp.makeConstraints { make in
```
Comment (nbhasin2) [inline Client/Frontend/Widgets/CellWithRoundedButton.swift:59]: personal nit: move away from snapkit

File: Client/Frontend/Widgets/CellWithRoundedButton.swift (hunk 3 above)
Comment (nbhasin2) [reply Client/Frontend/Widgets/CellWithRoundedButton.swift:20]: keeping this as using `.build` type it in closures strangely gives no input

**Pull request merged**